lib/api/core/entrypoints/embedded/protocols/websocket.js

```diff
@@ -264,79 +327,124 @@ class WebsocketProtocol extends Protocol {
-      this.channels[channel].count--;
-    }
-    else {
-      delete this.channels[channel];
-    }
+    ids.delete(connectionId);
 
-    const index = connection.channels.indexOf(channel);
-    if (index !== -1) {
-      connection.channels.splice(index, 1);
+    if (ids.size === 0) {
+      this.channels.delete(channel);
     }
+
+    connection.channels.delete(channel);
   }
 
   disconnect(clientId, message = 'Connection closed by remote host') {
     debug('[%s] disconnect', clientId);
 
-    if (this.connectionPool[clientId]) {
-      this.connectionPool[clientId].socket.close(1011, message);
+    const connection = this.connectionPool.get(clientId);
+    if (connection) {
+      connection.alive = false;
+      connection.socket.close(1011, message);
     }
   }
 
   _send (id, data) {
     debug('[%s] send: %a', id, data);
 
-    const connection = this.connectionPool[id];
+    const connection = this.connectionPool.get(id);
 
-    if (connection && connection.alive && connection.socket.readyState === connection.socket.OPEN) {
+    if (connection && connection.alive &&
+      connection.socket.readyState === connection.socket.OPEN
+    ) {
       connection.socket.send(data);
     }
   }
 
+  _startHeartbeat(heartbeat) {
+    assert(
+      Number.isInteger(heartbeat) && heartbeat >= 0,
+      `WebSocket: invalid heartbeat value ${heartbeat}`);
+
+    if (heartbeat > 0) {
+      this.heartbeatTimeout = heartbeat;
+      this.heartbeatInterval = setInterval(
+        this._doHeartbeat.bind(this),
+        heartbeat);
+    }
+  }
+
+  _startIdleSweeps(idleTimeout) {
+    assert(
+      Number.isInteger(idleTimeout) && idleTimeout >= 0,
+      `WebSocket: invalid idleTimeout value ${idleTimeout}`);
+
+    if (idleTimeout > 0) {
+      this.idleTimeout = idleTimeout;
+      this.idleSweepInterval = setInterval(
+        this._sweepIdleSockets.bind(this),
+        this.idleTimeout);
+    }
+  }
+
   _doHeartbeat() {
-    debug('[WebSocket] Heartbeat');
-    for (const id of Object.keys(this.connectionPool)) {
-      // did not respond since we last sent a PING request
-      if (this.connectionPool[id].alive === false) {
+    debug('Heartbeat');
+    const
+      lastActivityThreshold = this.activityTimestamp - this.heartbeatTimeout;
+
+    for (const connection of this.connectionPool.values()) {
+      if (connection.alive === false) {
+        // if still marked as "not alive", then the socket did not respond
+        // to the last emitted PING request
+        // (this correctly triggers the 'close' event handler on that socket)
+        connection.socket.terminate();
+      } else if (connection.lastActivity < lastActivityThreshold) {
+        // emit a PING request only if the socket has been inactive for longer
+        // than the heartbeat value
+        connection.alive = false;
+        connection.socket.ping();
+      }
+    }
+  }
+
+  _sweepIdleSockets() {
+    debug('Cleaning dead sockets');
+
+    const now = Date.now();
```

Comment:
Why you do not use `Date.now()` directly ? Or `this.activityTimestamp` ?

Reply:
I'm getting Date.now here and not inside the loop, because we're scanning ALL sockets registered in the connections pool, and that may mean hundred of thousands Date.now invocations saved. The performance gain here is substantial.

Also, I decided to not use `activityTimestamp` because, as the name states, it's an activity timestamp, so it seems wrong to use it in that context. Especially considering that it would save only 1 Date.now invocation per `idleTimeout` (which will probably be a delay of a few minutes minimum), and that really isn't much.